server/swap/swap.go

```diff
@@ -874,19 +880,30 @@ func bufferedTicker(ctx context.Context, dur time.Duration) chan struct{} {
 }
 
 func (s *Swapper) tryConfirmSwap(ctx context.Context, status *swapStatus, confTime time.Time) (final bool) {
-	status.mtx.Lock()
-	defer status.mtx.Unlock()
-	if status.swapTime.IsZero() || !status.swapConfirmed.IsZero() {
-		return
+	status.mtx.RLock()
+	if status.swapTime.IsZero() {
+		status.mtx.RUnlock()
+		return // no swap yet to confirm
+	}
+	if !status.swapConfirmed.IsZero() {
+		status.mtx.RUnlock()
+		return true // already confirmed
 	}
+	status.mtx.RUnlock() // Contract.Confirmations can timeout (slow), causing lock contention
+
 	confs, err := status.swap.Confirmations(ctx)
```

Comment:
That's interesting.  We certainly don't want 3700 init requests for the same match all to start coin waiters.

Regarding the `status.swap` field, we could also disallow setting it if it's not nil, since it doesn't make sense to change it once it's set.  At least that was my thought process in doing this - The `swapTime.IsZero` check is what needs to be under lock, with the assumption that `status.swap` would never be modified after being initially set.  Yeah, I think that needs to be explicitly prevented.

I'll certainly look at this more closely.

Reply:
I think you identified a more fundamental and nasty issue.  I've tried to address that with https://github.com/decred/dcrdex/pull/1563/commits/4d82ea4ba00b8f80e29e42451856bd97c54f1205